[Woo POS] [Variable Products] Add parent product to the POSItem enum, and make UI

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14700
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This pull request adds support for showing the parent product card for a variable product in the POS item selector. The main changes include adding a new `ParentProductCardView`, creating reusable view modifiers for item card styles and a reusable image view to share UI in the simple and parent product card views.

New components and views:

* [`WooCommerce/Classes/POS/Presentation/ParentProductCardView.swift`](diffhunk://#diff-210e84ea8e59459811d01a2b6ef2dd5536513ef37ecded0eaff994597f7d1267R1-R81): Added a new `ParentProductCardView` struct to display parent products in the POS system. This view includes a detailed view for variable products. When supporting other product types with a parent product (e.g. bundle), the detailed view can be configured based on the parent product type.
* [`WooCommerce/Classes/POS/Presentation/Reusable Views/POSItemImageView.swift`](diffhunk://#diff-6622145f20f58fa1cf7d0fc7c922796db46e93cfe364190f42029be226937a6fR1-R34): Introduced a new `POSItemImageView` struct to handle the display of product images within POS item cards. I was wondering if we could just use a placeholder image in `ProductImageThumbnail`, but didn't want to change the pre-existing behavior until we're working on UI/UX polishes.

Refactoring and reusable styles:

* [`WooCommerce/Classes/POS/Presentation/Reusable Views/POSItemCardBorderStylesModifier.swift`](diffhunk://#diff-9911ceecd39f4e08e91981d632212a39da1c1c5b47e9bfeb3f156693140574e2R1-R29): Created a new `POSItemCardBorderStylesModifier` struct to apply consistent border styles to item cards. This modifier is used in both `SimpleProductCardView` and `ParentProductCardView`.
* [`WooCommerce/Classes/POS/Presentation/SimpleProductCardView.swift`](diffhunk://#diff-68c880ea22cd0d896c93b09d412b85170905af695bc79820f03b9ab7c303f693L16-L30): Refactored `SimpleProductCardView` to use the new `POSItemImageView` and `POSItemCardBorderStylesModifier` for improved code reusability and consistency. [[1]](diffhunk://#diff-68c880ea22cd0d896c93b09d412b85170905af695bc79820f03b9ab7c303f693L16-L30) [[2]](diffhunk://#diff-68c880ea22cd0d896c93b09d412b85170905af695bc79820f03b9ab7c303f693L51-L67)

These changes collectively improve the maintainability and visual consistency of the POS product card views.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

🗒️ As the variable products support require WC version 9.6+ which is not publicly released at the time of writing, I'd recommend commenting out [L214 in ProductsRemote](https://github.com/woocommerce/woocommerce-ios/blob/efe8572e3c69a04a66bc76d2e41b855bb5ff02aa/Networking/Networking/Remote/ProductsRemote.swift#L214) to include all product types.

- Switch to a store with at least one variable product
- Go to Menu > Point of Sale --> scroll down the list, and make sure the variable products show up in the list without a price and with "Options available" below the product name (no changes expected in the simple product card UI to minimize the risk of this PR)

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Compared the product card UI with trunk to ensure there are no UI changes in the simple product cards.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

dark | light
-- | --
![Simulator Screenshot - iPad Pro 11-inch (M4) - 2024-12-31 at 16 36 45](https://github.com/user-attachments/assets/08d54868-3a92-47b2-8b6f-09a3431f6ac9) | ![Simulator Screenshot - iPad Pro 11-inch (M4) - 2024-12-31 at 16 38 02](https://github.com/user-attachments/assets/c43e0879-0315-40c2-9f52-2bfbe1bb79eb)


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

jaclync committed:
Add `parentProduct: POSParentProduct` case to `POSItem` enum, parse products based on product type in item service, and show new `ParentProductCardView` for parent products.

jaclync committed:
Reuse border styles and image view in both simple and parent product card views.

jaclync committed:
Revert changes on the simple product card UI.

jaclync committed:
ParentProductCardView: polishes.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WooCommerce/Classes/POS/Presentation/SimpleProductCardView.swift:
```diff
@@ -13,21 +13,11 @@ struct SimpleProductCardView: View {
+                             imageSize: Constants.productCardSize * scale,
+                             scale: scale)
+            .frame(width: min(Constants.productCardSize * scale, Constants.maximumProductCardSize),
+                   height: Constants.productCardSize * scale)
```
```suggestion
                   height: Constants.productCardSize * scale)
               .clipped()
```

I think we need to re-add clipped to limit the growth of images for extremely large accessibility sizes:

| With Clipped | Without Clipped |
|-------------|-----------------|
| ![with clipped](https://github.com/user-attachments/assets/66d68cfb-8a02-48df-9917-08fe2690ec74) | ![without clipped](https://github.com/user-attachments/assets/27d189a2-60b5-45cd-bcf9-1d4f74f3f518) |

For `ParentProductCardView` as well.

jaclync replied:
Great catch, `clipped` needs to be placed after the `frame`. I think it'd be great to remove the if/else condition in the product image view and use a placeholder image instead of setting up the view differently based on the image source.

jaclync replied:
IMO the ideal UX is to scale to fit with the same aspect ratio, I checked a few product images and they are cropped too much. I compared it with trunk, the cropping behavior is the same but it's out of the scope of this PR. I created a separate issue for this: https://github.com/woocommerce/woocommerce-ios/issues/14776.

<img src="https://github.com/user-attachments/assets/857384c1-2f53-4969-aa8d-685a5a965c06" width="400" />

jaclync replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/14771/commits/093b9520275b549afbd3d0c4af7a053bead2860d.

staskus replied:
> I created a separate issue for this: https://github.com/woocommerce/woocommerce-ios/issues/14776.

Thank you!

staskus commented on WooCommerce/Classes/POS/Presentation/ParentProductCardView.swift:
```diff
@@ -0,0 +1,81 @@
+                             imageSize: Constants.productCardSize * scale,
+                             scale: scale)
+            .frame(width: min(Constants.productCardSize * scale, Constants.maximumProductCardSize),
+                   height: Constants.productCardSize * scale)
```
```suggestion
                   height: Constants.productCardSize * scale)
              .clipped()
```

jaclync replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/14771/commits/093b9520275b549afbd3d0c4af7a053bead2860d.

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -170,6 +170,12 @@ private extension ItemListView {
             }
+        case .parentProduct(let parentProduct):
+            Button {
+                print("tapped parent product")
```
Since we now treat POS as released, we should ensure that these half-baked states don't get into production. 

I know that for now variable products only appear for unreleased Woo versions but maybe we should hide them behind the feature flag to feel safe during the development. What are your thoughts on this?

jaclync replied:
Yes, variable products are only loaded remotely when the local feature flag is enabled (disabled in production):

https://github.com/woocommerce/woocommerce-ios/blob/7c42ba8fd15211f43b9c0e93ab4a3fbd64acbed4/Yosemite/Yosemite/PointOfSale/PointOfSaleProductService.swift#L44-L46

Therefore, variable parent products do not show up in the item selector in POS regardless of their WC version.

staskus replied:
Ah, great!

staskus commented on WooCommerce/Classes/POS/Utils/Color+WooCommercePOS.swift:
```diff
@@ -87,6 +87,17 @@ extension Color {
 
+    // MARK: - POS Items
+
+    static var posItemSecondaryText: Color {
```
I wonder if we should have a separate color only for the "Options Available" text. We cleaned up after M1 to limit the number of colors and have a consistent POS color palette. I see we started reintroducing more view-specific colors. I suppose the reason is new designs that do not follow a consistent color palette. Did you take this color from Figma?

IMHO, ideally we would have primary/secondary/tertiary text and background colors which would cover most if not all the POS. For some special scenarios, we would define colors within specific views.

This comment is not a blocker, although it would be great if we reuse the existing color and clarify if this particular color is intended to be reused anywhere else.

jaclync replied:
Yes, this gray shade 60 is from the design rNAclWFAdQRw7EXEcsNVai-fi-2653_11376. It makes sense to reuse the colors from the POS color constants. I also had to guess the dark mode version, so we will want a design review anyways. Updated in https://github.com/woocommerce/woocommerce-ios/pull/14771/commits/d68a7f09e10cdf3ac9d3f792ef474f87fe66c273.

staskus reviewed (approved):
Thanks for the work!

I'm approving. Good job on making certain parts reusable while splitting the parent and simple product use-cases which are likely to keep diverging in the future 👍 .

Before merging:

- Address the large accessibility sizes issue

For this / future tasks:

- Make sure we don't leave POS in a half-working state during the project development
- Try to reuse existing colors / clarify if "Options available" is supposed to be a special case in the POS color pallete.

Thanks!

jaclync committed:
Remove one-off color for parent product card secondary text.

jaclync committed:
Move `clipped` to be after `frame` for it to work in large font sizes.

jaclync commented:
Thanks for catching the large font size issue and the feedback on reusing the colors @staskus! I believe I've addressed the comments and updated the PR. The screenshots in the PR description are also up to date with the existing POS secondary text color, which looks good in both light and dark modes.
